src/Psl/DateTime/DateTime.php

```diff
@@ -0,0 +1,490 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\DateTime;
+
+use IntlCalendar;
+use Psl\Locale\Locale;
+
+/**
+ * @immutable
+ *
+ * @psalm-suppress ImpureMethodCall
+ */
+final readonly class DateTime implements DateTimeInterface
+{
+    use DateTimeConvenienceMethodsTrait;
+
+    private Timezone $timezone;
+
+    private Timestamp $timestamp;
+
+    private int $year;
+
+    /**
+     * @var int<1, 12>
+     */
+    private int $month;
+
+    /**
+     * @var int<1, 31>
+     */
+    private int $day;
+
+    /**
+     * @var int<0, 23>
+     */
+    private int $hours;
+
+    /**
+     * @var int<0, 59>
+     */
+    private int $minutes;
+
+    /**
+     * @var int<0, 59>
+     */
+    private int $seconds;
+
+    /**
+     * @var int<0, 999999999>
+     */
+    private int $nanoseconds;
+
+    /**
+     * Constructs a new date-time instance with specified components and timezone.
+     *
+     * This constructor initializes a date-time object with the provided year, month, day, hour, minute,
+     * second, and nanosecond components within the given timezone. It ensures that all components are within their
+     * valid ranges: nanoseconds [0, 999,999,999], seconds [0, 59], minutes [0, 59], hours [0, 23], month [1, 12],
+     * and day [1, 28-31] depending on the month and leap year status. The constructor validates these components and
+     * assigns them to the instance if they are valid. If any component is out of its valid range, an
+     * `Exception\InvalidDateTimeException` is thrown.
+     *
+     * @throws Exception\InvalidArgumentException If any of the date or time components are outside their valid ranges,
+     *                                            indicating an invalid date-time configuration.
+     *
+     * @mutation-free
+     */
+    private function __construct(Timezone $timezone, Timestamp $timestamp, int $year, int $month, int $day, int $hours, int $minutes, int $seconds, int $nanoseconds)
+    {
+        if (
+            $nanoseconds < 0 || $nanoseconds >= NANOSECONDS_PER_SECOND ||
+            $seconds < 0 || $seconds >= 60 ||
+            $minutes < 0 || $minutes >= 60 ||
+            $hours < 0 || $hours >= 24 ||
+            $month < 1 || $month > 12 ||
+            $day < 1 || $day > 31 || $day > Month::from($month)->getDaysForYear($year)
+        ) {
+            throw new Exception\InvalidArgumentException('One or more components of the date-time are out of valid ranges.');
```

Comment:
This error message is not very helpful though.

Reply:
agree, we also need to use `OutOfRange` exception instead of InvalidArgumentException here.

will split this check into multiple to throw more clear messages.